File: redirect/click-aggregator-api/src/dto/clickstream_dto.rs
use serde::{Deserialize, Serialize};
use salvo::oapi::ToSchema;
use chrono::{DateTime, Utc};
use crate::model::clickstream::ClickStreamItem;

/// Data Transfer Object for ClickStreamItem API responses
#[derive(Clone, Serialize, Deserialize, Debug, ToSchema)]
pub struct ClickStreamItemDto {
    /// Unique identifier for the click stream item
    pub id: String,
    /// Owner of the route
    pub owner_id: String,
    /// Creator of the route
    pub creator_id: String,
    /// Route identifier
    pub route_id: String,
    /// Workspace identifier
    pub workspace_id: String,
    /// Timestamp when the click occurred
    pub created: DateTime<Utc>,
    /// Destination URL
    pub dest: String,
    /// IP address of the clicker
    pub ip: String,
    /// Geographic continent (optional)
    pub continent: Option<String>,
    /// Geographic country (optional)
    pub country: Option<String>,
    /// Geographic location (optional)
    pub location: Option<String>,
    /// Operating system family (optional)
    pub os_family: Option<String>,
    /// Operating system version (optional)
    pub os_version: Option<String>,
    /// User agent family (optional)
    pub user_agent_family: Option<String>,
    /// User agent version (optional)
    pub user_agent_version: Option<String>,
    /// Device brand (optional)
    pub device_brand: Option<String>,
    /// Device family (optional)
    pub device_family: Option<String>,
    /// Device model (optional)
    pub device_model: Option<String>,
    /// First session timestamp (optional)
    pub session_first: Option<DateTime<Utc>>,
    /// Number of clicks in session (optional)
    pub session_clicks: Option<u64>,
    /// Whether this is a unique click
    pub is_unique: bool,
    /// Whether this click is from a bot
    pub is_bot: bool,
}

/// Response wrapper for click stream queries
#[derive(Clone, Debug, Serialize, Deserialize, ToSchema)]
pub struct ClickStreamResponseDto {
    /// List of click stream items
    pub items: Vec<ClickStreamItemDto>,
    /// Total count of items (for pagination)
    pub total: u64,
    /// Current page offset
    pub offset: u32,
    /// Current page limit
    pub limit: u32,
    /// Whether there are more pages
    pub has_more: bool,
}

impl From<ClickStreamItem> for ClickStreamItemDto {
    fn from(item: ClickStreamItem) -> Self {
        use crate::model::clickstream::{UNKNOWN, epoch_datetime};

        // Convert "_unknown" values to None for API responses
        let to_option = |s: String| {
            if s == UNKNOWN { None } else { Some(s) }
        };

        Self {
            id: item.id,
            owner_id: item.owner_id,
            creator_id: item.creator_id,
            route_id: item.route_id,
            workspace_id: item.workspace_id,
            created: item.created,
            dest: item.dest,
            ip: item.ip,
            continent: to_option(item.continent),
            country: to_option(item.country),
            location: to_option(item.location),
            os_family: to_option(item.os_family),
            os_version: to_option(item.os_version),
            user_agent_family: to_option(item.user_agent_family),
            user_agent_version: to_option(item.user_agent_version),
            device_brand: to_option(item.device_brand),
            device_family: to_option(item.device_family),
            device_model: to_option(item.device_model),
            session_first: if item.session_first == epoch_datetime() { None } else { Some(item.session_first) },
            session_clicks: if item.session_clicks == 0 { None } else { Some(item.session_clicks) },
            is_unique: item.is_unique,
            is_bot: item.is_bot,
        }
    }
}

impl From<&ClickStreamItem> for ClickStreamItemDto {
    fn from(item: &ClickStreamItem) -> Self {
        use crate::model::clickstream::{UNKNOWN, epoch_datetime};

        // Convert "_unknown" values to None for API responses
        let to_option = |s: &String| {
            if s == UNKNOWN { None } else { Some(s.clone()) }
        };

        Self {
            id: item.id.clone(),
            owner_id: item.owner_id.clone(),
            creator_id: item.creator_id.clone(),
            route_id: item.route_id.clone(),
            workspace_id: item.workspace_id.clone(),
            created: item.created,
            dest: item.dest.clone(),
            ip: item.ip.clone(),
            continent: to_option(&item.continent),
            country: to_option(&item.country),
            location: to_option(&item.location),
            os_family: to_option(&item.os_family),
            os_version: to_option(&item.os_version),
            user_agent_family: to_option(&item.user_agent_family),
            user_agent_version: to_option(&item.user_agent_version),
            device_brand: to_option(&item.device_brand),
            device_family: to_option(&item.device_family),
            device_model: to_option(&item.device_model),
            session_first: if item.session_first == epoch_datetime() { None } else { Some(item.session_first) },
            session_clicks: if item.session_clicks == 0 { None } else { Some(item.session_clicks) },
            is_unique: item.is_unique,
            is_bot: item.is_bot,
        }
    }
}

File: api/Application/DTOs/ClickStreamDto.cs
namespace ShortasProxyApi.Application.DTOs;

/// <summary>
/// DTO for click stream data matching the non-nullable ClickHouse schema.
/// Uses default values instead of nulls: '_unknown' for strings, epoch for DateTime, 0 for numbers.
/// </summary>
public class ClickStreamDto
{
    /// <summary>
    /// Constant representing unknown/missing data
    /// </summary>
    public const string Unknown = "_unknown";

    /// <summary>
    /// Epoch timestamp representing unknown/missing session data
    /// </summary>
    public static readonly DateTime EpochDateTime = new DateTime(1970, 1, 1, 0, 0, 0, DateTimeKind.Utc);

    public string Id { get; set; } = string.Empty;
    public string OwnerId { get; set; } = string.Empty;
    public string CreatorId { get; set; } = string.Empty;
    public string RouteId { get; set; } = string.Empty;

    /// <summary>Route name (enriched from local data)</summary>
    public string? RouteName { get; set; }

    /// <summary>Route domain name (enriched from local data)</summary>
    public string? RouteDomainName { get; set; }

    public string WorkspaceId { get; set; } = string.Empty;
    public DateTime Created { get; set; }
    public string Dest { get; set; } = string.Empty;
    public string Ip { get; set; } = string.Empty;

    /// <summary>Geographic continent (defaults to "_unknown")</summary>
    public string Continent { get; set; } = Unknown;

    /// <summary>Geographic country (defaults to "_unknown")</summary>
    public string Country { get; set; } = Unknown;

    /// <summary>Geographic location (defaults to "_unknown")</summary>
    public string Location { get; set; } = Unknown;

    /// <summary>Operating system family (defaults to "_unknown")</summary>
    public string OsFamily { get; set; } = Unknown;

    /// <summary>Operating system version (defaults to "_unknown")</summary>
    public string OsVersion { get; set; } = Unknown;

    /// <summary>User agent family (defaults to "_unknown")</summary>
    public string UserAgentFamily { get; set; } = Unknown;

    /// <summary>User agent version (defaults to "_unknown")</summary>
    public string UserAgentVersion { get; set; } = Unknown;

    /// <summary>Device brand (defaults to "_unknown")</summary>
    public string DeviceBrand { get; set; } = Unknown;

    /// <summary>Device family (defaults to "_unknown")</summary>
    public string DeviceFamily { get; set; } = Unknown;

    /// <summary>Device model (defaults to "_unknown")</summary>
    public string DeviceModel { get; set; } = Unknown;

    /// <summary>First session timestamp (defaults to epoch: 1970-01-01)</summary>
    public DateTime SessionFirst { get; set; } = EpochDateTime;

    /// <summary>Number of clicks in session (defaults to 0)</summary>
    public long SessionClicks { get; set; } = 0;

    public bool IsUnique { get; set; }
    public bool IsBot { get; set; }

    /// <summary>
    /// Check if a string field contains unknown/missing data
    /// </summary>
    public static bool IsUnknown(string? value) => string.IsNullOrEmpty(value) || value == Unknown;

    /// <summary>
    /// Check if this item has valid session data
    /// </summary>
    public bool HasSession() => SessionFirst != EpochDateTime;

    /// <summary>
    /// Check if this item has valid geographic data
    /// </summary>
    public bool HasGeoData() => !IsUnknown(Country);

    /// <summary>
    /// Check if this item has valid device data
    /// </summary>
    public bool HasDeviceData() => !IsUnknown(DeviceFamily);

    /// <summary>
    /// Check if this item has valid user agent data
    /// </summary>
    public bool HasUserAgentData() => !IsUnknown(UserAgentFamily);
}

